src/convert_image/avx2.rs

```diff
@@ -840,6 +841,147 @@ unsafe fn lrgb_to_i420_avx2(
+    _last_src_plane: u32,
+    src_strides: &[usize],
+    src_buffers: &[&[u8]],
+    _last_dst_plane: u32,
+    dst_strides: &[usize],
+    dst_buffers: &mut [&mut [u8]],
+) -> bool {
+    if width == 0 || height == 0 {
+        return true;
+    }
+
+    if dst_buffers.is_empty()
+        || dst_strides.is_empty()
+        || src_buffers.is_empty()
+        || src_strides.is_empty()
+    {
+        return false;
+    }
+
+    let max_stride = usize::max_value() / height as usize;
+    if (src_strides[0] > max_stride) || (dst_strides[0] > max_stride) {
+        return false;
+    }
+
+    if src_strides[0] * height as usize > src_buffers[0].len()
+        || dst_strides[0] * height as usize > dst_buffers[0].len()
+    {
+        return false;
+    }
+
+    const ALPHAS_MASK: [u8; 32] = [
+        0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00,
+        0xff, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00,
+        0x00, 0xff,
+    ];
+    const SHUFFLE_MASK: [i8; 32] = [
+        2, 1, 0, -1, 5, 4, 3, -1, 8, 7, 6, -1, 11, 10, 9, -1, 2, 1, 0, -1, 5, 4, 3, -1, 8, 7, 6,
+        -1, 11, 10, 9, -1,
+    ];
+
+    const ITEMS_PER_ITERATION: usize = 32;
+    const OUTPUT_BPP: usize = 4;
+    const INPUT_BPP: usize = 3;
+
+    let w = width as usize;
+    let output_stride_diff = if dst_strides[0] == 0 {
+        0
+    } else {
+        dst_strides[0] - (OUTPUT_BPP * w)
+    };
+    let input_stride_diff = if src_strides[0] == 0 {
+        0
+    } else {
+        src_strides[0] - (INPUT_BPP * w)
+    };
+
+    let mask_alphas = _mm256_loadu_si256(ALPHAS_MASK.as_ptr() as *const __m256i);
+    let mask_shuffle = _mm256_loadu_si256(SHUFFLE_MASK.as_ptr() as *const __m256i);
+    let output_buffer = dst_buffers[0].as_mut_ptr();
+    let input_buffer = src_buffers[0].as_ptr();
+    let mut ibuffer_offset = 0;
+    let mut obuffer_offset = 0;
+
+    for _ in 0..height {
+        _mm_prefetch(input_buffer as *const i8, _MM_HINT_NTA);
+
+        for _ in (0..width).step_by(ITEMS_PER_ITERATION) {
+            let input = _mm256_loadu2_m128i(
+                input_buffer.add(ibuffer_offset + 12) as *const __m128i,
+                input_buffer.add(ibuffer_offset) as *const __m128i,
+            );
+
+            let input1 = _mm256_loadu2_m128i(
+                input_buffer.add(ibuffer_offset + 24 + 12) as *const __m128i,
+                input_buffer.add(ibuffer_offset + 24) as *const __m128i,
+            );
+
+            let input2 = _mm256_loadu2_m128i(
+                input_buffer.add(ibuffer_offset + 48 + 12) as *const __m128i,
+                input_buffer.add(ibuffer_offset + 48) as *const __m128i,
+            );
+
+            let input3 = _mm256_loadu2_m128i(
+                input_buffer.add(ibuffer_offset + 72 + 12 - 4) as *const __m128i,
+                input_buffer.add(ibuffer_offset + 72) as *const __m128i,
+            );
+
+            let input3 =
```

Comment:
Could you please add some comment about why this schema is not symmetric?

You can attach this schema:
```
// bFgFrFbE gErEbDgD rDbCgCrC bBgBrBbA gArAb9g9 r9b8g8r8 b7g7r7b6 g6r6b5g5 r5b4g4r4 b3g3r3b2 g2r2b1g1 r1b0g0r0 bFgFrFbE gErEbDgD rDbCgCrC bBgBrBbA gArAb9g9 r9b8g8r8 b7g7r7b6 g6r6b5g5 r5b4g4r4 b3g3r3b2 g2r2b1g1 r1b0g0r0
//                                                                                                                                                                                     ^----------------0----------------^
//                                                                                                                                                          ^---------------12----------------^
//                                                                                                                               ^---------------24----------------^
//                                                                                                    ^---------------36----------------^
//                                                                         ^---------------48----------------^
//                                              ^---------------60----------------^
//                   ^---------------72----------------^
// ^---------------80----------------^
```

From this picture one can understand that load of the last group is not done the same way for the other groups because otherwise we risk to read outside the image bounds.